### PR TITLE
Ibex example Arty A7-100T

### PR DESCRIPTION
**Pull request by towoe**
Simple example of how Ibex can be used to run software on a FPGA target.

### CODE REVIEW

File: examples/fpga/artya7-100/rtl/top.sv [hunk 1]
```diff
@@ -0,0 +1,122 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+module top (
```
Comment (imphil) [inline examples/fpga/artya7-100/rtl/top.sv]: Can we find a more descriptive name than `top`? Once we start adding more examples, having multiple files named `top` in our repository makes it harder to communicate and to find them in an editor/IDE.

File: examples/fpga/artya7-100/rtl/top.sv [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+  /*
+  *  SRAM block for instruction and data storage
+  *
+  *  The address range of the core is move to the
+  *  storage address space.
+  */
+  ram_2p #(
+    .Width(32),
+    .Depth(4 * 1024 / 4) // 4 kB
```
Comment (imphil) [inline examples/fpga/artya7-100/rtl/top.sv]: Is that the maximum amount of blockram we can get on this board? Otherwise I'd try to go for 64kB.

File: examples/fpga/artya7-100/rtl/top.sv [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+  ) u_ram (
+    .clk_a_i     (clk_sys),
+    .clk_b_i     (clk_sys),
+
+    .a_req_i     (instr_req),
+    .a_write_i   ('b0),
+    .a_addr_i    ((instr_addr - 32'h00001000)/4), // Translate addr to storage
```
Comment (imphil) [inline examples/fpga/artya7-100/rtl/top.sv]: Let's do this like a normal interconnect would do it:
- Gate the request signal with the address mask for the RAM area
- The RAM byte address is then only the LSB parts of the address.
- Please assign the result of these operations to an intermediate signal instead of doing the logic inside the ram_2p instantiation. This way you can have signal names which "speak for themselves", e.g. "instr_addr_index" (with index being a typical name for word addresses; without this suffix one would assume byte addresses)

**Comment by towoe** [comment]
Thanks a lot @imphil for the comments! I will incorporate those soon.

File: examples/fpga/artya7-100/rtl/top.sv (hunk 1 above)
Comment (towoe) [reply examples/fpga/artya7-100/rtl/top.sv]: do we want something like `top_artya7_100` or maybe `top_alpine_artya7_100` to have a little more Ibex in there?

File: examples/fpga/artya7-100/rtl/top.sv (hunk 2 above)
Comment (towoe) [reply examples/fpga/artya7-100/rtl/top.sv]: I was aiming more for the minimum, but I see that it might be preferable to go for a bigger default.

File: examples/fpga/artya7-100/rtl/top.sv (hunk 1 above)
Comment (imphil) [reply examples/fpga/artya7-100/rtl/top.sv]: I'd go for something like top_artya7_100, and use the same name in the core file (=> "lowrisc:ibex:top_artya7_100")

File: examples/README.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+This will create a directory `build` which contains the output files, including
+the bitstream.
+
+## Program
+
+Open the Vivado project:
+`build/lowrisc_ibex_top_artya7-100_0.1/synth-vivado/lowrisc_ibex_top_artya7-100_0.1.xpr`
+
+And use the Hardware Manager to program the Arty board:
+`build/lowrisc_ibex_top_artya7-100_0.1/synth-vivado/lowrisc_ibex_top_artya7-100_0.1.bit`
```
[the remark this answers is not in the document]
Comment (towoe) [reply examples/README.md]: done

File: examples/fpga/artya7-100/rtl/clkgen_xil7series.sv [hunk 5]
```diff
@@ -0,0 +1,92 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+module clkgen_xil7series (
+    input IO_CLK,
+    input IO_RST_N,
+    output clk_sys,
+    output clk_48MHz,
```
[the remark this answers is not in the document]
Comment (towoe) [reply examples/fpga/artya7-100/rtl/clkgen_xil7series.sv]: done

File: examples/fpga/artya7-100/rtl/clkgen_xil7series.sv [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+  );
+
+  BUFG clk_48_bufg (
+    .I (clk_48_unbuf),
+    .O (clk_48_buf)
+  );
+
+  // outputs
+  // clock
+  assign clk_sys = clk_50_buf; // TODO: choose 50 MHz clock as sysclock for now
```
[the remark this answers is not in the document]
Comment (towoe) [reply examples/fpga/artya7-100/rtl/clkgen_xil7series.sv]: I was able to get something around 57 MHz (1200/21). Could you maybe test if for example 60 MHz is not possible?

File: examples/fpga/artya7-100/rtl/top.sv [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+     .fetch_enable_i        ('b1)
+  );
+
+  /*
+  *  SRAM block for instruction and data storage
+  *
+  *  The address range of the core is move to the
+  *  storage address space.
+  */
```
[the remark this answers is not in the document]
Comment (towoe) [reply examples/fpga/artya7-100/rtl/top.sv]: done

File: examples/fpga/artya7-100/rtl/top.sv [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    .b_req_i     (data_req),
+    .b_write_i   (data_we),
+    .b_addr_i    ((data_addr - 32'h00001000)/4), // Translate addr to storage
+    .b_wdata_i   (data_wdata),
+    .b_rdata_o   (data_rdata)
+  );
+
+  /*
+  * Connect the led output the the lower four bits of a written data word
+  */
```
[the remark this answers is not in the document]
Comment (towoe) [reply examples/fpga/artya7-100/rtl/top.sv]: done

File: examples/fpga/artya7-100/rtl/top.sv [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+  logic [3:0] leds;
+  always_ff @(posedge clk_sys or negedge rst_sys_n) begin
+    if (!rst_sys_n) begin
+      leds <= 4'b0;
+    end else if (data_we) begin
+      leds <= data_wdata[3:0];
+    end else begin
+      leds <= leds;
```
[the remark this answers is not in the document]
Comment (towoe) [reply examples/fpga/artya7-100/rtl/top.sv]: done

File: examples/fpga/artya7-100/rtl/top.sv [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+  );
+
+  /*
+  * Connect the led output the the lower four bits of a written data word
+  */
+  logic [3:0] leds;
+  always_ff @(posedge clk_sys or negedge rst_sys_n) begin
+    if (!rst_sys_n) begin
+      leds <= 4'b0;
+    end else if (data_we) begin
```
[the remark this answers is not in the document]
Comment (towoe) [reply examples/fpga/artya7-100/rtl/top.sv]: done

File: examples/sw/led/Makefile [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+endif
+ARCH = rv32imc
+# ARCH = rv32im # to disable compressed instructions
+SRCS = $(PROGRAM).c
+
+CC = /tools/riscv/bin/riscv32-unknown-elf-gcc
+
+OBJCOPY ?= $(subst gcc,objcopy,$(wordlist 1,1,$(CC)))
+OBJDUMP ?= $(subst gcc,objdump,$(wordlist 1,1,$(CC)))
+REGTOOL = ../../util/regtool.py
```
[the remark this answers is not in the document]
Comment (towoe) [reply examples/sw/led/Makefile]: removed

File: examples/sw/led/led.c [hunk 12]
```diff
@@ -0,0 +1,47 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdint.h>
+#define CLK_FIXED_FREQ_HZ (50ULL * 1000 * 1000)
```
[the remark this answers is not in the document]
Comment (towoe) [reply examples/sw/led/led.c]: checked, thanks for the pointer!

File: examples/fpga/artya7-100/rtl/top.sv (hunk 1 above)
Comment (towoe) [reply examples/fpga/artya7-100/rtl/top.sv]: used top_artya7_100

File: examples/fpga/artya7-100/rtl/top.sv (hunk 2 above)
Comment (towoe) [reply examples/fpga/artya7-100/rtl/top.sv]: I used 64kB with 48kB for the "rom" and 16kB for data

File: examples/fpga/artya7-100/rtl/top.sv (hunk 3 above)
Comment (towoe) [reply examples/fpga/artya7-100/rtl/top.sv]: Thanks for the guidance, could you check if this is ok now?

File: examples/fpga/artya7-100/rtl/top.sv (hunk 2 above)
Comment (albmoriconi) [reply examples/fpga/artya7-100/rtl/top.sv]: I can see how both choices (maximal vs minimal amount) have their own merit; however, wouldn't clocking multiple (unused) BRAMs add to the device power consumption?

Comment (imphil) [reply examples/fpga/artya7-100/rtl/top.sv]: That's certainly possible, but low power consumption is a not a goal for this simple demo example, helping people to get started with Ibex.

File: examples/fpga/artya7-100/rtl/top_artya7_100.sv [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+     .instr_gnt_i           ('b1),
+     .instr_rvalid_i        ('b1),
+     .instr_addr_o          (instr_addr),
+     .instr_rdata_i         (instr_rdata),
+
+     .data_req_o            (data_req),
+     .data_gnt_i            ('b1),
+     .data_rvalid_i         ('b1),
+     .data_we_o             (data_we),
+     .data_be_o             (data_be),
```
Comment (albmoriconi) [inline examples/fpga/artya7-100/rtl/top_artya7_100.sv]: The data byte-write enable is connected to the core out, but not used. Could this be a problem if another application is used?

Comment (towoe) [reply examples/fpga/artya7-100/rtl/top_artya7_100.sv]: For simplicity I am ignoring the byte enable, as the SRAM is only addressed on a word level basis.
What kind of problem are you thinking about?

Comment (towoe) [reply examples/fpga/artya7-100/rtl/top_artya7_100.sv]: And following the comment from @imphil the interface to the SRAM is going to be changed.

Comment (albmoriconi) [reply examples/fpga/artya7-100/rtl/top_artya7_100.sv]: I guess it would be used in case of a misaligned write; sure @imphil can shed some light.

Comment (towoe) [reply examples/fpga/artya7-100/rtl/top_artya7_100.sv]: Yes, you are right, misaligned accesses use the data enable signal. And ignoring this is not correct if one accesses the data storage on a byte granularity and or not word aligned. But the purpose of this example is to have a simple starting point and not a complete system, as the direct connection of the LEDs to the write data signal is also an example of.

**Comment by towoe** [comment]
>     * Can we go for a unified address space for data and instruction, starting from 0? (So essentially remove the ROM/RAM masking.)
Changed to 0
>     * Can we go from a 2p memory to a 1p memory? The core won't issue a data and instruction operation at the same time anyways, so we can go simpler here (in line with having that a minimal example). You need to have a small arbitration block between the two interfaces (data/instruction) in front of the SRAM (essentially a mini-bus). (Please ping me on Zulip if that paragraph doesn't make sense to you.)
Using 1p memory now
>     * Optional: 57 MHz looks rather strange. So I'd go back to 50 MHz if we cannot go higher to a number that looks "normal". (But I leave that up to you, changing things around all the time isn't the greatest use of time.)
I only put in 57 MHz to show the highest one possible for now, but I agree, 50 MHz is nicer, so used this now again.

**Comment by jrrk** [comment]
> The core won't issue a data and instruction operation at the same time anyways, so we can go simpler here (in line with having that a minimal example). 

@imphil This seems to be a convenient but surprising limitation. Why go for two buses at all at the top level of the Ibex if the two buses are never used simultaneously ?

File: examples/.gitignore [hunk 14]
```diff
@@ -0,0 +1 @@
+build
```
Comment (imphil) [inline examples/.gitignore]: As in the tracer example: please use fusesoc from the source root, and put this gitignore there as well.

File: doc/examples.rst [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+Examples
+========
+
+To make use of Ibex it has to be integrated as described in :ref:`core-integration`.
+Here is an example of how this could look like.
+
+FPGA
+----
+
+A minimal example for the `Arty A7 <https://reference.digilentinc.com/reference/programmable-logic/arty-a7/start>`_ FPGA Development board is provided. Find the description of how to use this in ``examples/fpga/artya7-100/README.md``.
```
Comment (imphil) [inline doc/examples.rst]: I'd actually prefer having the documentation in the manual and link to it from the README, i.e. the other way around.

File: examples/fpga/artya7-100/README.md [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+
+This is a simple example of how Ibex can be utilized.
+The core is not connected to a bus, but to the SRAM directly, which is initialized
+by the `led.vmem` file from the software section.
+
+## Requirements
+
+### Tools
+
+  - RISC-V compiler
```
Comment (imphil) [inline examples/fpga/artya7-100/README.md]: A RV32 compiler

File: examples/fpga/artya7-100/README.md [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+  - Xilinx Vivado
+
+## Build
+
+### Software
+
+First the software must be built. Go into `sw/led` and call:
+
+```
+make CC=/path/to/RISC-V-compiler
```
Comment (imphil) [inline examples/fpga/artya7-100/README.md]: Can you make this `/path/to/riscv32-unknown-elf-gcc` to make it easier to find the right compiler for people?

File: examples/fpga/artya7-100/README.md [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+This will create a directory `build` which contains the output files, including
+the bitstream.
+
+## Program
+
+Open the Vivado project:
+`build/lowrisc_ibex_top_artya7_100_0.1/synth-vivado/lowrisc_ibex_top_artya7_100_0.1.xpr`
+
+And use the Hardware Manager to program the Arty board:
+`build/lowrisc_ibex_top_artya7_100_0.1/synth-vivado/lowrisc_ibex_top_artya7_100_0.1.bit`
```
Comment (imphil) [inline examples/fpga/artya7-100/README.md]: `fusesoc pgm` should do the trick as well

File: examples/fpga/artya7-100/rtl/top_artya7_100.sv [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+    output [3:0]        LED
+);
+
+  parameter int          MEM_SIZE  = 64 * 1024; // 64 kB
+  parameter logic [31:0] MEM_START = 32'h00000000;
+  parameter logic [31:0] MEM_MASK  = MEM_SIZE-1;
+
+  logic clk_sys, rst_sys_n;
+
+  // Instruction connection so SRAM
```
Comment (imphil) [inline examples/fpga/artya7-100/rtl/top_artya7_100.sv]: so -> to

File: examples/fpga/artya7-100/rtl/top_artya7_100.sv [hunk 20]
```diff
@@ -0,0 +1,154 @@
...
+  logic        instr_req;
+  logic        instr_gnt;
+  logic        instr_rvalid;
+  logic [31:0] instr_addr;
+  logic [31:0] instr_rdata;
+
+  // Data connection so SRAM
```
Comment (imphil) [inline examples/fpga/artya7-100/rtl/top_artya7_100.sv]: so -> to

File: examples/fpga/artya7-100/rtl/top_artya7_100.sv [hunk 21]
```diff
@@ -0,0 +1,154 @@
...
+  logic        mem_req;
+  logic        mem_write;
+  logic [31:0] mem_wdata;
+  logic        mem_rvalid;
+  logic [31:0] mem_rdata;
+
+
+  ibex_core #(
+     .RV32E(0),
+     .RV32M(1),
```
Comment (imphil) [inline examples/fpga/artya7-100/rtl/top_artya7_100.sv]: Don't set these two parameters, it's good to test the default configuration (whatever that might be if we change things)

File: examples/fpga/artya7-100/rtl/top_artya7_100.sv [hunk 22]
```diff
@@ -0,0 +1,154 @@
...
+  end
+  assign LED = leds;
+
+  // Clock and reset
+  clkgen_xil7series
+    clkgen(
+      .IO_CLK       (IO_CLK),
+      .IO_RST_N     (IO_RST_N),
+      .clk_sys      (clk_sys),
+      .rst_sys_n    (rst_sys_n)
```
Comment (imphil) [inline examples/fpga/artya7-100/rtl/top_artya7_100.sv]: If you prefer, you can simplify this to:

```

  clkgen_xil7series
    clkgen(
      .IO_CLK
      .IO_RST_N 
      .clk_sys
      .rst_sys_n
  );
```

File: examples/sw/led/Makefile [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+$(PROGRAM).elf: $(OBJS) $(LINKER_SCRIPT)
+	$(CC) $(CFLAGS) -T $(LINKER_SCRIPT) $(OBJS) -o $@ $(LIBS)
+
+%.dis: %.elf
+	$(OBJDUMP) -SD $^ > $@
+
+# Note: this target requires the srecord package to be installed.
+# XXX: This could be replaced by objcopy once
+# https://sourceware.org/bugzilla/show_bug.cgi?id=19921
+# is merged.
```
Comment (imphil) [inline examples/sw/led/Makefile]: This bug is already merged now, but I'm not sure if the implementation has been cleaned up enough to be usable. And it's not really available in a standard toolchain yet, so slighly reword this to read "is merged" -> "is widely available"

File: examples/fpga/artya7-100/README.md (hunk 17 above)
Comment (asb) [reply examples/fpga/artya7-100/README.md]: That can also cause confusion unfortunately - multilib builds will use riscv64-unknown-elf-gcc but may include libs for ilp32.

File: examples/fpga/artya7-100/rtl/top_artya7_100.sv (hunk 19 above)
Comment (towoe) [reply examples/fpga/artya7-100/rtl/top_artya7_100.sv]: done

File: examples/fpga/artya7-100/rtl/top_artya7_100.sv (hunk 20 above)
Comment (towoe) [reply examples/fpga/artya7-100/rtl/top_artya7_100.sv]: done

File: examples/fpga/artya7-100/rtl/top_artya7_100.sv (hunk 21 above)
Comment (towoe) [reply examples/fpga/artya7-100/rtl/top_artya7_100.sv]: done

File: examples/fpga/artya7-100/rtl/top_artya7_100.sv (hunk 22 above)
Comment (towoe) [reply examples/fpga/artya7-100/rtl/top_artya7_100.sv]: Yes, this looks better. Done.

File: examples/sw/led/Makefile (hunk 23 above)
Comment (towoe) [reply examples/sw/led/Makefile]: Good to know. I guess this would also simplify `link.ld` in regards to the alignment of the sections.

File: examples/fpga/artya7-100/README.md (hunk 17 above)
Comment (towoe) [reply examples/fpga/artya7-100/README.md]: So `/path/to/riscvXX-unknown-elf-gcc`?

**Comment by towoe** [comment]
I will make the changes to the documentation after the tracer PR is merged as this simplifies the rebase.

File: examples/fpga/artya7-100/README.md (hunk 17 above)
Comment (imphil) [reply examples/fpga/artya7-100/README.md]: Well, even the *unknown* part is not really standard (unfortunately), with the rather commonly used [Eclipse prebuilts](https://gnu-mcu-eclipse.github.io/toolchain/riscv/install/) defaulting to *none* here. I'd go for either riscv32-unknown-elf-gcc (which is what you get with a SiFive toolchain source build) or just leave it as you had it.

File: doc/examples.rst (hunk 15 above)
Comment (towoe) [reply doc/examples.rst]: So the description should move to doc, but the README will still hold the build instruction?

File: examples/fpga/artya7-100/README.md (hunk 17 above)
Comment (towoe) [reply examples/fpga/artya7-100/README.md]: I just checked the Eclipse site, looks like they now use `riscv-none-embed-gcc`.
I will just put in some examples of what this path could look like. I would assume that the chances are high that if someone finds the way to Ibex Arty example, they are able to set the path to the right compiler correctly.

File: examples/fpga/artya7-100/README.md (hunk 18 above)
Comment (towoe) [reply examples/fpga/artya7-100/README.md]: Thanks for that, much better to program this way.

**Comment by imphil** [comment]
Looks like this is ready to be merged now?

One more thing: This currently doesn't support byte reads/writes. That's fine for now, but we should at least add a TODO in the code and file an issue for it.

**Comment by towoe** [comment]
I will add a TODO regarding the byte access.
My last push was already rebased on https://github.com/lowRISC/ibex/pull/77 so the commits are appearing here as well now, so we should wait until tracer is merged so I can rebase on the master branch for better attribution.

**Comment by imphil** [comment]
This should be ready after a rebase on top of master now that the tracer has gone in.

**Comment by imphil** [comment]
Thanks Tobias!
